### ISSUE TITLE
NPE while fetching certain indexes during code generation

### ISSUE THREAD

**Issue by axelfontaine** [jOOQ/jOOQ#7127]
### Expected behavior and actual behavior:
Expected: java file generation as it was working with 3.9

Actual:
```
java.lang.NullPointerException
        at org.jooq.util.AbstractIndexDefinition.<init>(AbstractIndexDefinition.java:53)
        at org.jooq.util.postgres.PostgresDatabase$1.<init>(PostgresDatabase.java:193)
        at org.jooq.util.postgres.PostgresDatabase.getIndexes0(PostgresDatabase.java:193)
        at org.jooq.util.AbstractDatabase.getIndexes(AbstractDatabase.java:1734)
        at org.jooq.util.AbstractDatabase.getIndexes(AbstractDatabase.java:1767)
        at org.jooq.util.AbstractTableDefinition.getIndexes(AbstractTableDefinition.java:87)
        at org.jooq.util.JavaGenerator.generateTable(JavaGenerator.java:3692)
        at org.jooq.util.JavaGenerator.generateTable(JavaGenerator.java:3527)
        at org.jooq.util.JavaGenerator.generateTables(JavaGenerator.java:3514)
        at org.jooq.util.JavaGenerator.generate(JavaGenerator.java:451)
        at org.jooq.util.JavaGenerator.generate(JavaGenerator.java:405)
        at org.jooq.util.JavaGenerator.generate(JavaGenerator.java:327)
        at org.jooq.util.GenerationTool.run(GenerationTool.java:642)
        at org.jooq.util.GenerationTool.generate(GenerationTool.java:203)
        at org.jooq.util.maven.Plugin.execute(Plugin.java:192)
        at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo(DefaultBuildPluginManager.java:134)
        at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:208)
        at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:154)
        at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:146)
        at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject(LifecycleModuleBuilder.java:117)
        at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject(LifecycleModuleBuilder.java:81)
        at org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build(SingleThreadedBuilder.java:51)
        at org.apache.maven.lifecycle.internal.LifecycleStarter.execute(LifecycleStarter.java:128)
        at org.apache.maven.DefaultMaven.doExecute(DefaultMaven.java:309)
        at org.apache.maven.DefaultMaven.doExecute(DefaultMaven.java:194)
        at org.apache.maven.DefaultMaven.execute(DefaultMaven.java:107)
        at org.apache.maven.cli.MavenCli.execute(MavenCli.java:993)
        at org.apache.maven.cli.MavenCli.doMain(MavenCli.java:345)
        at org.apache.maven.cli.MavenCli.main(MavenCli.java:191)
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:498)
        at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced(Launcher.java:289)
        at org.codehaus.plexus.classworlds.launcher.Launcher.launch(Launcher.java:229)
        at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode(Launcher.java:415)
        at org.codehaus.plexus.classworlds.launcher.Launcher.main(Launcher.java:356)
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:498)
        at org.apache.maven.wrapper.BootstrapMainStarter.start(BootstrapMainStarter.java:39)
        at org.apache.maven.wrapper.WrapperExecutor.execute(WrapperExecutor.java:122)
        at org.apache.maven.wrapper.MavenWrapperMain.main(MavenWrapperMain.java:50)```
```

This is preceded by other messages like

```
Ignoring foreign key     :my_table__my_table_col1_fkey (myschema.my_table.col1) referencing table_pk references a schema out of scope for jooq-meta: null
```

Interestingly other things like Records and Table classes are generated just fine.

Related to #6624 

### Versions:

- jOOQ: 3.10.4
- Java: 8u162
- Database (include vendor): PostgreSQL 9.4
- JDBC Driver (include name if inofficial driver): PostgreSQL 42.0.0

**Comment by lukaseder** [comment]
What crazy index edge-case situation have you created? :) Do you know how to reproduce this with a minimal, complete, verifiable example? Are there indexes in one schema referencing tables in other schemas and one of the schemas is excluded from code generation?

**Comment by axelfontaine** [comment]
All objects live within a single schema. Nothing exotic.

**Comment by lukaseder** [comment]
I see, thanks for the update. So, definitely some problem where one object is included and the other one isn't. This shouldn't result in any exception, of course. Will investigate.

**Comment by lukaseder** [comment]
I've tried reproducing this on 3.11.0-SNAPSHOT and 3.10.4 as follows:

```sql
create table t_7127 (
  include_me INT,
  exclude_me INT
);
create index i_7127_a on t_7127 (include_me, exclude_me);
create index i_7127_b on t_7127 (include_me) where (exclude_me is null);
create unique index i_7127_c on t_7127 (include_me) where (exclude_me is null);
```

And then, roughly:

```xml
<includes>.*</includes>
<excludes>exclude_me</excludes>
<includeExcludeColumns>true</includeExcludeColumns>
```

Unfortunately with no success.

> With nspname populated for all rows.

Yes it's unlikely not to be populated, but perhaps it is excluded? Thus my question if there is any cross-schema index reference or something like that.

Anyway, there should be a guard for `null` schemas. Even if I cannot reproduce it as you've described so far, in the event of cross-schema index references where one schema is excluded, we shouldn't instantiate the `IndexDefinition`.

**Comment by axelfontaine** [comment]
As I said, just one schema (but not `public` if that helps...) and therefore no cross-schema references.

**Comment by lukaseder** [comment]
Oh, I see it now. The query looking for indexes is not filtering the `nspname` column for input schemata. This way, all the indexes from your database (including those in the catalog) will be fetched in the query, which results in that NPE.

The current query is:

```sql
select 
  "tnsp"."nspname", 
  "trel"."relname", 
  "irel"."relname", 
  "i"."indisunique", 
  array(select pg_get_indexdef("i"."indexrelid", k + 1, true)
  from generate_subscripts("i"."indkey", 1) as k
  order by k) as "columns", 
  "i"."indoption"::int[] as "asc_or_desc"
from "pg_catalog"."pg_index" as "i"
  join "pg_catalog"."pg_class" as "irel"
  on "irel".oid = "i"."indexrelid"
  join "pg_catalog"."pg_class" as "trel"
  on "trel".oid = "i"."indrelid"
  join "pg_catalog"."pg_namespace" as "tnsp"
  on "tnsp".oid = "trel"."relnamespace"
order by 
  1, 
  2, 
  3
```

And it should be:

```sql
select 
  "tnsp"."nspname", 
  "trel"."relname", 
  "irel"."relname", 
  "i"."indisunique", 
  array(select pg_get_indexdef("i"."indexrelid", k + 1, true)
  from generate_subscripts("i"."indkey", 1) as k
  order by k) as "columns", 
  "i"."indoption"::int[] as "asc_or_desc"
from "pg_catalog"."pg_index" as "i"
  join "pg_catalog"."pg_class" as "irel"
  on "irel".oid = "i"."indexrelid"
  join "pg_catalog"."pg_class" as "trel"
  on "trel".oid = "i"."indrelid"
  join "pg_catalog"."pg_namespace" as "tnsp"
  on "tnsp".oid = "trel"."relnamespace"
where "tnsp"."nspname" in (?, ?, ..., ?) -- Missing predicate here
order by 
  1, 
  2, 
  3
```

**Comment by lukaseder** [comment]
... there might be some edge case (that I still cannot reproduce), which leads to the schema being excluded, but the table (a wrong table?) being found nevertheless.

**Comment by lukaseder** [comment]
> What could also be a possible cause: the primary keys are UUIDs and they have a default value obtained using the uuid_generate_v4() function which is automatically placed by the uuid-ossp extension in the public schema...

I don't think it would cause this particular NPE. The code generator currently doesn't resolve the default expressions but generates them as "plain SQL".

**Comment by lukaseder** [comment]
This is now fixed and I'll merge it to 3.10.4. Since I couldn't reproduce the NPE, would you be interested in giving short feedback on the improved logic? You can find the current `PostgresDatabase` implementation with the fix here:

https://github.com/jOOQ/jOOQ/blob/master/jOOQ-meta/src/main/java/org/jooq/util/postgres/PostgresDatabase.java#L150
